### ISSUE TITLE
report unknown context types in "block" definitions

### ISSUE THREAD

**Issue by faxm0dem** [balabit/syslog-ng#1803]
Tested on syslog-ng-3.13.2

```
block root set-unless-empty(
  template()
  value()
)
{
  set(
    "`template`"
    value(`value`)
    condition(
      not '' eq "`template`"
    )
  );
};

rewrite r_all {
    set-unless-empty(
        template(
            "${SEQNUM}"
        ),
        value(
            syslog.seqnum
        )
    );
 }
```

This results in the following error message on configuration parsing:

```
Error parsing rewrite expression, rewrite plugin set-unless-empty not found in /etc/syslog-ng/syslog-ng.conf at line 75, column 5:

    set-unless-empty(
    ^^^^^^^^^^^^^^^^
```

**Comment by faxm0dem** [comment]
The documentation states the following:

> The root blocks can be used in the "root" context of the configuration file, that is, outside any other statements

So the change seems consistent with the documentation, but it breaks my config which should not have worked in the first place :-)

Do you have a suggestion on how to change my config?

**Comment by faxm0dem** [comment]
Reading the doc again, I see I should set the type to `rewrite`. Now it works.
leaving this for reference here, and closing.

Sorry for the noise :-)

**Comment by bazsi** [comment]
I see this just resolved itself :) just one note: we never knowingly
intended to change this, and the root context should have only worked on
the top-level, so it was a bug that it worked elsewhere as well.

-- 
Bazsi

**Comment by faxm0dem** [comment]
yes I agree, it was a bug. I guess this issue will help others who misread the config like me ;)

**Comment by faxm0dem** [comment]
@bazsi sorry I misread my initial configuration. In fact I used the following block instead:

```
block set set-unless-empty(
  template()
  value()
)
{
  set(
    "`template`"
    value(`value`)
    condition(
      not '' eq "`template`"
    )
  );
};
```

So I actually used a `set` block.
Before this used to work in a rewrite block, and now it doesn't.

I guess the right behaviour should be for the configuration parser to fail if block type isn't one of `destination, filter, log, parser, rewrite, root, source`
